index.js

```diff
@@ -120,6 +120,45 @@ const Method = (prefix) => ({
   }
 });
 
+const PassthroughFunction = (lambda, prefix) => ({
+  Type: 'AWS::Lambda::Function',
+  Properties: {
+    Code: {
+      ZipFile: cf.join('\n', [
+        '"use strict";',
+        'const AWS = require("aws-sdk");',
+        cf.sub('const lambda = new AWS.Lambda({ region: "${AWS::Region}" });'),
+        'module.exports.lambda = (event, context, callback) => {',
+        '  if (event.httpMethod === "OPTIONS") {',
+        '    const response = {',
+        '      statusCode: 200,',
+        '      "Access-Control-Allow-Headers": "Content-Type"',
+        '      "Access-Control-Allow-Methods": "POST, OPTIONS"',
+        '      "Access-Control-Allow-Origin": "*"',
+        '    };',
+        '    callback(null, response);',
+        '    return;',
+        '  }',
+        `  const lambdaParams = { FunctionName: "${lambda}" };`,
+        '  lambda.invoke(lambdaParams, (err, response) => {',
+        '    if (err) return callback(err);',
+        '    if (!response) callback(new Error("Response required"));',
+        '    response.headers = response.headers || {};',
+        '    response.headers["Access-Control-Allow-Origin"] = "*";',
+        '    callback(null, response)',
+        '  });',
+        '};'
+      ])
+    },
+    Role: cf.getAtt(`${prefix}FunctionRole`, 'Arn'),
```

Comment:
All good points. For this one, though, I don't think I understand what's going on. `AWS::Lambda::Permission`s seem to exist on their own and are not in fact associated with a role. I could create a new one of those and change an existing one to get this:

```js
const PassthroughFunctionPermission = (lambda, prefix) => ({
  Type: 'AWS::Lambda::Permission',
  Properties: {
    FunctionName: cf.ref(`${prefix}PassthroughFunction`),
    Action: 'lambda:InvokeFunction',
    Principal: 'apigateway.amazonaws.com',
    SourceArn: cf.sub(`arn:aws:execute-api:\${AWS::Region}:\${AWS::AccountId}:\${${prefix}Api}/*`)
  }
});

const PassthroughPermission = (lambda, prefix) => ({
  Type: 'AWS::Lambda::Permission',
  Properties: {
    FunctionName: cf.ref(lambda),
    Action: 'lambda:InvokeFunction',
    Principal: 'lambda.amazonaws.com',
    SourceArn: cf.sub(`arn:aws:execute-api:\${AWS::Region}:\${AWS::AccountId}:\${${prefix}Api}/*`)
  }
});
```

But I don't know if that's right or how I would associate any of that with the role. The role is supposed allow (something) for the principal Lambda — is there something else it needs?

Reply:
`AWS::Lambda::Permission` is generally used to give some AWS service permission to invoke a lambda function. The first one, gives `apigateway.amazonaws.com` permission to invoke, as long as the invoker is the api-gateway identified by that source arn.

The second one looks wrong. It would give `lambda.amazonaws.com` permission to invoke, as long as the invoker is an api-gateway. That mismatch doesn't seem right.

The `AWS::IAM::Role` that is associated with a Lambda function is where you define _what that lambda function is allowed to do_. In this case, you want to add a statement to that role so that the Passthrough function can invoke the user's function. You shouldn't need an `AWS::Lambda::Permission` to do that.